### PR TITLE
Bell inequalities variational - Example

### PR DESCRIPTION
**Pull request by igres26**
Finding a maximally entangled state and the optimal measurement direction for maximal violation of CHSH inequalities.

- [x] Readme

### CODE REVIEW

**Comment by scarrazza** [comment]
@igres26 I will review this PR, it seems complete to me, could you please confirm?

**Comment by igres26** [comment]
I think it is complete yes. Thanks @scarrazza.

File: examples/bell-variational/main.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+def main(nshots):
+    '''Variationally find a maximally entangled state and the correct measurement angle
+       for violation of Bell inequalities.
+    Args:
+        nshots: number of shots to use for the minimization.
+        
+    '''
+    initial_parameters = np.random.uniform(0, 2*np.pi, 2)
+    circuits = set_parametrized_circuits()
+    best, params, extra = optimize(cost_function, initial_parameters, args=(circuits, nshots))
```
Comment (scarrazza) [inline examples/bell-variational/main.py:16]:
```suggestion
    best, params, _ = optimize(cost_function, initial_parameters, args=(circuits, nshots))
```

File: examples/bell-variational/main.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    print(f'Cost: {best}\n')
+    print(f'Parameters: {params}\n')
+    frequencies = []
+    for circuit in circuits:
+        circuit.set_parameters(params)
+        frequencies.append(circuit(nshots=nshots).frequencies())
+    chsh = compute_chsh(frequencies)
+    print(f'CHSH inequality value: {chsh}\n')
+    print(f'Target: {np.sqrt(2)*2}\n')
```
Comment (scarrazza) [inline examples/bell-variational/main.py:25]: We should probably include the relative difference between target and optimized value.

File: examples/bell-variational/main.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--nshots", default=10000, type=int)
```
Comment (scarrazza) [inline examples/bell-variational/main.py:30]:
```suggestion
    parser.add_argument("--nshots", default=10000, type=int, help='Number of shots for each circuit base')
```

File: examples/bell-variational/main.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    args = vars(parser.parse_args())
+    main(**args)
```
Comment (scarrazza) [inline examples/bell-variational/main.py:32]:
```suggestion
    main(**args)

```

File: examples/bell-variational/functions.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    chsh = 0
+    aux = 0
+    for freq in frequencies:
+        for outcome in freq:
+            if aux == 1:
+                chsh -= (-1)**(int(outcome[0])+int(outcome[1]))*freq[outcome]
+            else:
+                chsh += (-1)**(int(outcome[0])+int(outcome[1]))*freq[outcome]
+        aux += 1
+    nshots = sum(freq[x] for x in freq)
```
Comment (scarrazza) [inline examples/bell-variational/functions.py:60]: This line looks suspicious. This is using the last element of frequencies, freq is undefined in this context.

File: examples/bell-variational/README.md [hunk 6]
```diff
@@ -0,0 +1,49 @@
+# Maximal violation of Bell inequalities variationally
+
+Code at: [https://github.com/Quantum-TII/qibo/tree/master/examples/bell-variational](https://github.com/Quantum-TII/qibo/tree/master/examples/bell-variational)
+
+## Introduction
+
+During the advent of quantum mechanics, the properties of entanglement, or non-local correlations between quantum systems, were discussed or even doubted by prominent physicists. John Bell proposed an experiment where non-local interactions would be proven using correlations between measurements on two different qubits.
+
+This experiment has been performed and proven multiple times since. Now that quantum computers are avalilable, the prove of quantum non-locality is not only an interesting exercise but also an easy way to quickly gauge how "quantum" a device is.
```
Comment (scarrazza) [inline examples/bell-variational/README.md:9]: available

File: examples/bell-variational/README.md [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+This maximum violation only occurs on maximally entangled states and with a specific setting of the (a, a', b, b') parameters. These parameters are known, but the aim of this example is to variationally find the optimal parameters in order to fully violate this inequality.
+
+## Setup of the example
+
+This example consists of a very simple parametrized circuit with two variational angles. The first one sets up the state, and the second one sets up the direction of the measurements. Then the circuit is measured in all possible combinations of the Z and X basis.
+
+![circuit](images/circuit.png)
+
+Ideally, the circuit should prepare a maximally entangled state and set up a direction of measurment of 45 (225) degrees.
```
Comment (scarrazza) [inline examples/bell-variational/README.md:37]: measurement

I think we could print this angles in the example.

Comment (scarrazza) [reply examples/bell-variational/README.md:37]: ... in degrees

**Review by scarrazza — CHANGES_REQUESTED** [review]
@igres26 thank you very much for this nice and clean example.
Please find above some comments.

**Comment by scarrazza** [comment]
Concerning the sphinx docs, I think the equation image is maybe too big, please have a look.

File: examples/bell-variational/functions.py (hunk 5 above)
Comment (igres26) [reply examples/bell-variational/functions.py]: It was a way to compute the nshots without giving it to the function, but now added as a parameter.

File: examples/bell-variational/main.py (hunk 2 above)
Comment (igres26) [reply examples/bell-variational/main.py:26]: Added the relative distance

File: examples/bell-variational/README.md [hunk 8]
```diff
@@ -34,7 +34,7 @@ This example consists of a very simple parametrized circuit with two variational
 
 ![circuit](images/circuit.png)
 
-Ideally, the circuit should prepare a maximally entangled state and set up a direction of measurment of 45 (225) degrees.
+Ideally, the circuit should prepare a maximally entangled state and set up a direction of measurement of 45 (225) in degrees.
```
Comment (scarrazza) [inline examples/bell-variational/README.md:37]: Thanks, I see that the angles printed by the script are [87, 43], maybe we should say something about that here?

**Comment by scarrazza** [comment]
@igres26 thanks for the changes. Could you please include an extra argparse option for setting the backend?

Another point raised by JI, is the possibility to have a second backup circuit ansatz with 2 cnot gates, which may help reinforcing the entanglement.

**Comment by igres26** [comment]
@scarrazza I added the backend switcher. I am not sure if complicating the circuit and adding more trainable parameters is a good idea as a first example. Moreover, for maximal entanglement it just has to decide a 90 degree angle and it just works. But I can add some modifications to make the problem more complex.

**Comment by scarrazza** [comment]
@igres26 ok, I agree that we should start with a simple case, and for this tutorial the current circuit should be sufficient.

File: examples/bell-variational/main.py [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    chsh = compute_chsh(frequencies, nshots)
+    print(f'CHSH inequality value: {chsh}\n')
+    print(f'Target: {np.sqrt(2)*2}\n')
+    print(f'Relative distance: {100*np.abs(np.abs(chsh)-np.sqrt(2)*2)/np.sqrt(2)*2}%\n')
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--nshots", default=10000, type=int, help='Number of shots for each circuit base')
+    parser.add_argument("--backend", default='defaulteinsum', type=str, help='Backend to use for the example')
```
Comment (scarrazza) [inline examples/bell-variational/main.py:36]:
```suggestion
    parser.add_argument("--backend", default='custom', type=str, help='Backend to use for the example')
```

File: examples/bell-variational/main.py [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+import argparse
+
+
+def main(nshots, backend):
+    '''Variationally find a maximally entangled state and the correct measurement angle
+       for violation of Bell inequalities.
+    Args:
+        nshots: number of shots to use for the minimization.
+        backend: choice of backend to run the example in.
+
```
Comment (scarrazza) [inline examples/bell-variational/main.py:14]: Remove empty line.

File: examples/bell-variational/functions.py [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+def cost_function(parameters, circuits, nshots):
+    '''Compute the cost function that attempts to maximize the violation of CHSH inequalities.
+    Args:
+        parameters (np.array): parameters for the variational algorithm. First determines the
+                               initial state and second the measurement direction.
+        circuits (list(:class:`qibo.core.circuit.Circuit`)): Bell circuits in the 4 needed basis.
+        nshots (int): number of sampling for the experiment.
+    
+    Returns:
+        cost (real): closer to 0 as the inequalities are maximally violated. Can go negative due
```
Comment (scarrazza) [inline examples/bell-variational/functions.py:74]:
```suggestion
        cost (float): closer to 0 as the inequalities are maximally violated. Can go negative due
```

File: examples/bell-variational/functions.py [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+def compute_chsh(frequencies, nshots):
+    '''Computes the CHSH inequelity value given the restults of an experiment.
+    Args:
+        frequencies (list): list of dictionaries with the result for the measurement
+                            and the number of times it repeated, for each measurement 
+                            basis
+        nshots (int): number of samples taken in each run of the circuit.
+    
+    Returns:
+        chsh (real): value of the CHSH inequality.
```
Comment (scarrazza) [inline examples/bell-variational/functions.py:49]:
```suggestion
        chsh (float): value of the CHSH inequality.
```

**Review by scarrazza — COMMENTED** [review]
@igres26 thanks, I have place some small changes, but this is almost ready to merge.

**Comment by scarrazza** [comment]
@igres26 thanks, looks perfect!

**Review by scarrazza — APPROVED** [review]

**Pull request merged**